common/src/escape.rs

```diff
@@ -0,0 +1,246 @@
+        }
+    }
+
+    #[inline]
+    pub const fn to_char(&self) -> char {
+        match self {
+            Quote::Single => '\'',
+            Quote::Double => '"',
+        }
+    }
+}
+
+pub struct EscapeLayout {
+    pub quote: Quote,
+    pub len: Option<usize>,
+}
+
+pub trait Escape {
+    type Source: ?Sized;
+
+    fn source_len(&self) -> usize;
+    fn layout(&self) -> &EscapeLayout;
+
+    fn output_layout_with_checker(
+        source: &Self::Source,
+        preferred_quote: Quote,
+        reserved_len: usize,
+        length_add: impl Fn(usize, usize) -> Option<usize>,
+    ) -> EscapeLayout;
+    // fn output_layout(source: &Self::Source, preferred_quote: Quote) -> EscapeLayout {
+    //     Self::output_layout_with_checker(source, preferred_quote, 2, |a, b| a.checked_add(b))
+    // }
+    fn output_layout(source: &Self::Source, preferred_quote: Quote) -> EscapeLayout {
+        Self::output_layout_with_checker(source, preferred_quote, 2, |a, b| {
+            Some((a as isize).checked_add(b as isize)? as usize)
+        })
+    }
+
+    fn write_source(&self, formatter: &mut impl std::fmt::Write) -> std::fmt::Result;
+    fn write_body_slow(&self, formatter: &mut impl std::fmt::Write) -> std::fmt::Result;
+    fn write_body(&self, formatter: &mut impl std::fmt::Write) -> std::fmt::Result {
+        if self.layout().len == Some(self.source_len()) {
+            self.write_source(formatter)
+        } else {
+            self.write_body_slow(formatter)
+        }
+    }
+    fn write_quoted(&self, formatter: &mut impl std::fmt::Write) -> std::fmt::Result {
+        let quote = self.layout().quote.to_char();
+        formatter.write_char(quote)?;
+        self.write_body(formatter)?;
+        formatter.write_char(quote)
+    }
+    fn to_quoted_string(&self) -> Option<String> {
+        let len = self.layout().len?.checked_add(2)?;
+        let mut s = String::with_capacity(len);
+        self.write_quoted(&mut s).unwrap();
+        Some(s)
+    }
+}
+
+/// Returns the outer quotes to use and the number of quotes that need to be
+/// escaped.
+pub(crate) const fn choose_quote(
+    single_count: usize,
+    double_count: usize,
+    preferred_quote: Quote,
+) -> (Quote, usize) {
+    let (primary_count, secondary_count) = match preferred_quote {
+        Quote::Single => (single_count, double_count),
+        Quote::Double => (double_count, single_count),
+    };
+
+    // always use primary unless we have primary but no seconday
+    let use_secondary = primary_count > 0 && secondary_count == 0;
+    if use_secondary {
+        (preferred_quote.swap(), secondary_count)
+    } else {
+        (preferred_quote, primary_count)
+    }
+}
+
+pub struct UnicodeEscape<'a> {
+    source: &'a str,
+    layout: EscapeLayout,
+}
+
+impl<'a> UnicodeEscape<'a> {
+    pub fn with_forced_quote(source: &'a str, quote: Quote) -> Self {
+        let layout = EscapeLayout { quote, len: None };
+        Self { source, layout }
+    }
```

Comment:
Python requires length calculation due to allocating string object, but formatters doesn't.

will this be helpful for formatter?

```rust
let escape = if source.len() < TOO_LONG_TO_FORMAT {
    UnicodeEscape::output_layout(source, Quote::Double)
} else {
    UnicodeEscape::with_forced_quote(source, Quote::Double)
};

```